Pattern matching implementation

haxscramper opened the pull request:
Pattern matching implementation as described in https://github.com/nim-lang/RFCs/issues/245

WIP [article](https://gist.github.com/haxscramper/ad781ebd5a343b53197bf538a1252f83) on how to use this for writing macros.

haxscramper commented:
Right now use of `case` statement inside of generics fails, if identifiers like `_` are used in body. There are two workarounds right now: first one is to use `expand` macro that forces evaluation of the body, so you write things like

```nim
expand case a:
  of [_]: discard
  else: fail()
```

Second option is adding `mixin` statement for each unknown symbol, like this:

```nim
mixin _
case a:
  of [_]: discard
  else: fail()
```

This can be done automatically, but I don't know of a way to determine if a macro has been expanded inside of generic function or not (since use of `mixin` statement outside of procs/functions/methods is an error)

Yardanico commented on src/fusion/matching.rst:
```diff
@@ -0,0 +1,293 @@
+.. raw:: html
+  <blockquote><p>
+  "you can probably make a macro for that" -- Rika, 22-09-2020 10:41:51
```
@de-odex :))

juancarlospaco commented:
Needs changelog.

haxscramper commented:
Previous CI failure was related to C codegen differences between nim 1.0.0 and 1.2.0+ and happened only because I didn't account for it in tests. 

In addition to fixing this I also added support for `(field.subfield: <pattern>)` syntax - previously such construction had to be written using `(field: (subfield: <pattern>))` which was not really intuitive and quire bothersome to write. This addition is just purely for ergonomics/syntactic sugar.

I think that implementation is mostly feature-complete (almost all things described in RFC are covered, with some minor adjustments and several additions).

Things left are mostly related to better documentation and defensive programming of the DSL (test against more possible use cases & provide better error messages in case of malformed code).

When view types become none-experimental it might be possible to match objects using them.

haxscramper commented:
All fixes, additions are finally implemented, so now I need someone to review this.

I've significantly improved failed match exceptions messages, position for backtraces (using `{.line.}`), added better DSL compilation checks and added more tests. No new functionality was added.

haxscramper commented:
There were several things that I considered, but ultimately decided to not implement, or postpone until after the current version is accepted

### Support for capturing values into existing variables. 

It could be done by adding *yet another* syntax - `@=`, to explicitly declare that the existing variable should be reused. The another alternative is to check `declared` and then for `typeof` equality, but this is even less intuitive. So I decided it is not worth it, and you can just capture into a new variable and then `add`/`incl`/`=` to existing one however you like.

### Pragma annotations for variable captures

There is just not enough place in DSL, and honestly I don't think that this is something all that important, so I decided to not implement this.

### Using views for captured variables

Right now all captured variables introduce copies. I declare each variable at the top of the match expression, and then `=` or `add` to it during match evaluation. So matching `[until @vals == 5, .._] := @[2, 3, 4, 5]` would create `vals == @[2, 3, 4]]` sequence, where all elements will be copied. With view types (I suppose) it might be possible to declare `vals` as `seq[lent int]` and no copying would occur.

Because `views` are currently experimental this has not been implemented, but this is possibility for better performance, so it should not be overlooked. 

Another related problem - when `(maybeFuncMaybeField: [all < 10])` is captured, `maybeFuncMaybeField` is called multiple times, and this is not possible to determine whether `maybe..` is a field or a procedure *during pattern construciton*. I believe view types can also solve this problem, by doing

```nim
let vals: lent typeof(expr) = expr
```

and then using `vals` instead of an `expr` throughout the code. Without view types this can be done too, quite easily, just using 

```nim
let vals: typeof(expr) = expr
```

But it could (most likely will) create a lot of copies when `expr` is actually as field access. No I see this is a tradeoff between *calling `maybeFuncMaybeField` multiple times*, and *copying field* on each pattern check. Now there is no copying (since I assumed pattern matching will be done /mostly/ on fields, and things like `len` that are quite cheap to evaluate AFAIK).

haxscramper commented:
@timotheecour if you have time could you please review the implementation? Tests are currently failing due to CI issues, not implementation - everything is build-able & passes tests.

juancarlospaco commented on changelog.md:
```diff
@@ -1,5 +1,9 @@
 ## additions and changes
 
+- Pattern matching implementation as described in nim-lang/RFCs#245
```
```suggestion
- Add pattern matching implementation as described in nim-lang/RFCs#245
```

narimiran replied:
If we want to be pedantic, the changelog is about the changes that have happened, i.e. we should write "Add**ed**...".

haxscramper commented:
Is there anything that I can do that would speed up merge? I understand that reviewing four thousand lines of code is not a fast thing to do, but maybe I can at least make it easier somehow?

narimiran commented:
> Is there anything that I can do that would speed up merge?

Currently, releasing 1.4.2 is our priority and we're focusing on it.

If nothing happens with this for a week after 1.4.2 release, please ping us again.

juancarlospaco commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+    ## Set of all token-like nodes (primitive type literals or
+    ## identifiers)
+
+func codeFmt(str: string): string =
```
These are so tiny, it should be `template` or at least `{.inline.}`.

juancarlospaco commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+    of nnkOpenSymChoice: n[0].strVal()
+    of nnkSym: n.strVal()
+    of nnkStrKinds: n.strVal()
+    else: raiseAssert(&"Cannot get string value from node kind {n.kind}")
```
`case` wont really need indentation on the `of`, 
most of `case` on stdlib do not use this indentation, 
makes lines longer for no gain, like line 44, 
remove the indentation of the `of` IMHO.

juancarlospaco commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+  else:
+    raiseAssert(&"Cannot get variable from node kind {n.kind}")
+
+func firstDot(n: NimNode): NimNode =
```
Same, inline or template

juancarlospaco commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+      map:
+        shell
+
+    assert res is seq[string]
```
Awesome tests!, make some of them `static:` so no one breaks its ability to run at compile-time in the future ?.

timotheecour replied:
instead, use this pattern which is commonly used:
```nim
template main =
  # all the tests
main()
static: main()
```

haxscramper replied:
Changed use of `test` to `multitest` that runs code in both `static` and regular environment.

juancarlospaco commented:
LGTM

timotheecour commented on changelog.md:
```diff
@@ -1,5 +1,6 @@
 ## additions and changes
 
+- Added pattern matching implementation as described in nim-lang/RFCs#245
```
make sure this generates a clickable link

timotheecour commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+import std/[strutils, sequtils, strformat, sugar,
```
`tests/tmatching.nim` (following common convention)

timotheecour commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+  if a != b:
+    echo a
+    echo b
+    raiseAssert("Comparison failed in " & $instantiationInfo())
```
* don't use `echo`, the message should contain everything, you never know how the exception is handled.

```nim
raiseAssert("Comparison failed in " & $instantiationInfo() & "\nlhs: " & $a & "\nrhs: " & $b )
```

* why isn't `check a == b` sufficient?

haxscramper replied:
Using `check` is not possible in `static:` environment. Other suggestion noted & implemented

timotheecour commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+
+template fail(str: string): untyped =
+  echo "Fail on ", instantiationInfo(), ": ", str
+  fail()
```
ditto

timotheecour commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+          Infix   [@infixId, @lhs, @rhs]
+        ) := node
+
+        assert lhs is NimNode
```
use doAssert in tests, not assert, so it stays meaningful with `--assertions:off`

timotheecour commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+
+    main()
+
+    # block: [0 .. 2 @a is 12] := [12, 12, 12]; assert a == @[12, 12, 12]
```
`when false` to disable code so parser checks it

timotheecour commented on tests/tMatching.nim:
```diff
@@ -0,0 +1,1927 @@
+      assert nice == 69
+
+
+  when (1, 2, 0) <= (NimMajor, NimMinor, NimPatch):
```
more readable:
```nim
when (NimMajor, NimMinor, NimPatch) >= (1, 2, 0):
```
(or since pending https://github.com/nim-lang/fusion/issues/51)

timotheecour commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+    of nnkStrKinds: n.strVal()
+    else: raiseAssert(&"Cannot get string value from node kind {n.kind}")
+
+func lineIInfo(node: NimNode): NimNode =
```
lineIInfo ?

timotheecour commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+
+
+
+template `->`(a, b: bool): bool = (if a: b else: true)
```
not a common operator (and you only use it once), rename to a non-operator

timotheecour commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+template getSome[T](opt: Option[T], injected: untyped): bool =
+  opt.isSome() and ((let injected {.inject.} = opt.get(); true))
+
+func splitDots(n: NimNode): seq[NimNode] =
```
complex internal procs need doc comments

timotheecour commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2207 @@
+    drop[1..^1].foldl(nnkDotExpr.newTree(a, b))
+
+
+template assertKind(node: NimNode, kindSet: set[NimNodeKind]): untyped =
```
if these were regular procs you'd avoid the need for `instantiationInfo`

haxscramper replied:
This template is used as mostly implementation detail for validation of input AST and I find it much more convenient to have position in the source code alongside of expected node kinds, instead of looking over whole backtrace.

juancarlospaco commented:
Looks good still :+1:

haxscramper commented:
Any thoughts on the [article](https://gist.github.com/haxscramper/ad781ebd5a343b53197bf538a1252f83)? It mostly illustrates the main use case for the library - writing macros, but maybe I should add something else. I hope to put it on nim blog after this is merged and somewhat user-tested

juancarlospaco commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2208 @@
+    of nnkOpenSymChoice: n[0].strVal()
+    of nnkSym: n.strVal()
+    of nnkStrKinds: n.strVal()
+    else: raiseAssert(&"Cannot get string value from node kind {n.kind}")
```
Some of the code formats error messages using `strformat` and some uses `&` and `$`,
I do not know if theres a convention for this, maybe ask someone, I do not know if theres a reason,
but stdlib error messages are formatted using `&` and `$` AFAIK
:)

haxscramper replied:
Most likely that is done to avoid dependencies on strformat.

I used `$ &` in `quote` parts of codegen, to generate messages and strformat otherwise.

haxscramper commented:
@timotheecour is there anything else that needs fixing, or this looks good enough for you too?

haxscramper commented:
Added custom unpackers and changed sequence pattern matching to use `items` instead of random access via `[]`. 

This code doesn't work prior to `1.4.2.` and I have no idea as to why, so custom unpackers are only supported from `1.4.2` onwards.

```nim
type FieldIndex* = distinct int

template `[]`*(t: tuple, idx: static[FieldIndex]): untyped =
  t[idx.int]
  
echo (12, 12)[FieldIndex(0)]
```

juancarlospaco commented on src/fusion/matching.rst:
```diff
@@ -0,0 +1,418 @@
+  </p></blockquote>
+
+:Author: haxscramper
+:Version: 0.7
```
This is the version of the Spec ?, or version of what?, 
because individual files are not suppose to have individual versions and everything is part of Fusion version.

haxscramper replied:
This is a version spec, though I should probably remove it, because there is no significant meaning to it anyway.

juancarlospaco commented on tests/tmatching.nim:
```diff
@@ -0,0 +1,2078 @@
+      map:
+        shell
+
+    doAssert res is seq[string]
```
Tests are impressive!, 
maybe make at least some of them run at compile-time, just to be sure no one breaks your work in the future,
like the stuff that works at compile time make it run in a `static` block kinda like `static: tests()`, 
check how they are made on testament of Nim repo.

haxscramper replied:
They do get run at compile time - see # `multitest` and `multitestSince` templates.

Araq commented on src/fusion/matching.nim:
```diff
@@ -0,0 +1,2393 @@
+import sequtils, macros, tables, options, strformat, strutils,
+       parseutils, algorithm, hashes
+
+export options
```
In a follow-up PR, try to avoid this please.

Araq commented:
Merging this now but there will be followup changes to the documentation. This is still highly experimental. In the longer run I expect Nim to grow ``let`` / ``var`` keywords as expressions so that identifier injections become clearer. This means ``@val`` will become ``(let val)`` or similar. Happy to write an RFC for this.

haxscramper commented:
@Araq I planned to collect some feedback after this library is somewhat user-tested and update documentation & related parts accordingly.

About `let val` vs `@val` - I fully support this, it would look much better compared to current syntax.
